# Checking out new VCF inds
# Generate individual files for VCF_filtering

# Reads in VCF metadata and generates different individual files for VCF processing such as a list of outcrossing individuals that will not be filtered by heterozygosity

# arg[1] is the variant only vcf
# arg[2] is the invariant and variant all sites vcf
args = commandArgs(trailingOnly=TRUE)

# get ourput directory from environment
outdir <- Sys.getenv("OUTDIR")

# read in vcf metadata
vcf_info <- read.csv("~/capsella_sample_info/generated_mkwb/Capsella_vcf_metadata.txt",
                     header = T, sep = "\t")

##################### Variant only sites VCF
# read in list of samples in VCF generated by bcftools query -l <VCF>
# invcf <- read.csv("~/Documents/PhD/Research/vcf_filtering/individuals/allsites_cbp.txt",
#                   header = F, sep = "\t")

samples_var <- read.csv(arg[1], sep = "\t",
                        header = F)

# load dplyr
library(dplyr)
library(stringr)

# big data frame
dat <- left_join(samples_var, vcf_info, join_by("V1" == "vcf_sample_name"))

# create tibble for file
vcf_sp <- dat %>% group_by(species) %>% summarise(inds_pre = n())

# subset each species; just to know
cbp_all <- dat[which(dat$species == "Capsella bursa-pastoris"),]
cg <- dat[which(dat$species == "Capsella grandiflora"),]
cr <- dat[which(dat$species == "Capsella rubella"),]
co <- dat[which(dat$species == "Capsella orientalis"),]
np <- dat[which(dat$species == "Neslia paniculata"),]

### Not to be evaluated for odd heterozygosity
exp_het <- rbind(cg,np)

######### selecting only cbp I want--------
# remove F2s because I do not need those
cbp <- cbp_all[str_detect(cbp_all$sample_name, "F2_", negate = T),] #remove rows with F2 in the sample_name
# remove parents and pool
cbp <- cbp[str_detect(cbp$sample_name, "pool", negate = T),]
cbp <- cbp[str_detect(cbp$sample_name, "parent", negate = T),]

## get list of cbp I want to remove from vcf
rm <- cbp_all %>% anti_join(cbp)

# write removal list to file
write.table(rm$V1,
            paste(outdir, "removeCBP_vcf.txt", sep = "/"), 
            quote = F, sep = "\t", row.names = F, col.names = F)

# write C. grandiflora and Neslia paniculata to file so they are not removed during heterozygosity step
write.table(exp_het$V1, 
            paste(outdir, "CgNp.txt", sep = "/"), 
            quote = F, sep = "\t", row.names = F, col.names = F)


# generate table after removal of unwanted samples
# create tibble for file

sink("vcf_preprocessing.Rout")
dat %>% group_by(species) %>% summarise(inds_pre = n())

dat2 %>% group_by(species) %>% summarise(inds_post = n())
sink(NULL)

############## All Sites VCF

samples_allsites <- read.csv(arg[2], sep = "\t",
                        header = F)

dat <- left_join(samples_allsites, vcf_info, join_by("V1" == "vcf_sample_name"))
