# Checking out new VCF inds
# Generate individual files for VCF_filtering

# Reads in VCF metadata and generates different individual files for VCF processing such as a list of outcrossing individuals that will not be filtered by heterozygosity

# args[1] is the variant only vcf
# args[2] is the VCF prefix
# args[3] is the output directory
args <- commandArgs(trailingOnly=TRUE)

# get ourput directory from environment
outdir <- args[3]

##################### Variant only sites VCF
# read in vcf metadata
vcf_info <- read.csv("~/capsella_sample_info/generated_mkwb/Capsella_vcf_metadata.txt",
                     header = T, sep = "\t")

# read in list of samples in VCF generated by bcftools query -l <VCF>
samples_var <- read.csv(args[1], sep = "\t",
                        header = F)

# load dplyr
library(dplyr)
library(stringr)

# make data frame to rename JL new samples
samples_var$new <- str_remove_all(samples_var$V1, "_1.fastq.gz")

# join info of samples to samples in VCF
dat <- left_join(samples_var, vcf_info, join_by("new" == "vcf_sample_name"))

# create tibble for file
#vcf_sp <- dat %>% group_by(species) %>% summarise(inds_pre = n())

## subset each species; just to know
#cbp_all <- dat[which(dat$species == "Capsella bursa-pastoris"),]
#cg <- dat[which(dat$species == "Capsella grandiflora"),]
#cr <- dat[which(dat$species == "Capsella rubella"),]
#
########## selecting only cbp I want--------
## remove F2s because I do not need those
#cbp <- cbp_all[str_detect(cbp_all$sample_name, "F2_", negate = T),] #remove rows with F2 in the sample_name
## remove parents and pool
#cbp <- cbp[str_detect(cbp$sample_name, "pool", negate = T),]
#cbp <- cbp[str_detect(cbp$sample_name, "parent", negate = T),]
#
### get list of cbp I want to remove from vcf
#rm <- cbp_all %>% anti_join(cbp)
#
## write removal list to file
#write.table(rm$V1,
#            paste(outdir, "removeCBP_vcf.txt", sep = "/"), 
#            sep = "\t", row.names = F, col.names = F, quote = F)
#
## write C. grandiflora and Neslia paniculata to file so they are not removed during heterozygosity step
#write.table(cg$V1, 
#            paste(outdir, "CgNp.txt", sep = "/"), 
#            sep = "\t", row.names=F, col.names = F, quote = F)
#
# write name_change file for BCF tools
namefile <- paste0(args[2],"_new_names.txt")

write.table(dat[1:2], 
      file= paste(outdir, namefile, sep = "/"), 
      sep = "\t", row.names = F, col.names = F,, quote = F)
