#!/bin/bash
#
#SBATCH --job-name=CBPvariantVCFFilter
#SBATCH --nodes=10
#SBATCH --cpus-per-task=1
#SBATCH --ntasks-per-node=1
#SBATCH --time=0-50:00:00
#SBATCH --partition=josephsnodes
#SBATCH --account=josephsnodes
#SBATCH --mem-per-cpu=8G
#SBATCH --mail-type=ALL
#SBATCH --mail-user=wils1582@msu.edu
#SBATCH --output=/mnt/scratch/wils1582/slurm/slurm-%A_%a.out
# VCF filtering
# June 24, 2024
# Maya Wilson Brown
# Modified from C. Fiscus Capsella vcf filtering
# Additional modifications to associated R scripts
# The accompanying scripts will not output jpegs automatically
# Goals: 1) document how I have filtered the VCFs generated by Adrian
# 2) update software with new OS and newer versions (no VCFtools)
#
#
# change directory
cd /mnt/scratch/wils1582/
# purge modules
module purge
# load modules
module load BCFtools/1.18-GCC-12.3.0
module load PLINK/2.00a3.7-gfbf-2023a
module load R/4.3.2-gfbf-2023a

## vars
VCF=CBP_CRCG_rmCoF2.v.vcf.gz
FILEDIR=/mnt/home/wils1582/vcf_filtering/CBPCRCG_jul14
PREFIX=CBP_CRCG
RAWDIR=/mnt/research/josephslab/Maya/capsella/vcf/adrian_vcf/unfiltered
OUTDIR=/mnt/research/josephslab/Maya/capsella/vcf/adrian_vcf/final_filtered

#### PIPELINE #####
# remove C orientalis that snuck in and F2s
bcftools view --samples-file ^"$RAWDIR"/remove_CoF2.txt "$RAWDIR"/July14.final_called.v.vcf -Oz -o CBP_CRCG_rmCoF2.v.vcf.gz
# initial filters
# 
## GATK best practices hard filters
bcftools filter -e'QD < 2 | FS > 60 | SOR > 3 | MQ < 40 | MQRankSum < -12.5 | ReadPosRankSum < -8.0' \
	"$VCF" \
	> "$PREFIX"_temp.vcf
bcftools view -f.,PASS "$PREFIX"_temp.vcf \
	-Oz -o "$PREFIX"_filter1.vcf.gz

echo "Filter1 complete"

# start fitering report
touch "$FILEDIR"/log.txt

echo "$PREFIX" >> "$FILEDIR"/log.txt
echo "GATK best practices filter" >> "$FILEDIR"/log.txt
echo $(bcftools query -f'%CHROM %POS\n' "$PREFIX"_filter1.vcf.gz | wc -l) \
	>> "$FILEDIR"/log.txt
echo "Sample count" >> "$FILEDIR"/log.txt
echo $(bcftools query -l "$PREFIX"_filter1.vcf.gz | wc -l) \
	>> "$FILEDIR"/log.txt

# remove temp
rm "$PREFIX"_temp.vcf

## require 3 reads to call and keep only biallelic sites; dump entirely missing sites
bcftools filter -e'FMT/DP<3' -S . "$PREFIX"_filter1.vcf.gz | bcftools view -i 'F_MISSING<1' -m2 -M2 -Oz -o "$PREFIX"_temp2.vcf

# log progress
echo "temp2 complete"

echo "3 reads called and biallelic sites" >> "$FILEDIR"/log.txt
echo $(bcftools query -f'%CHROM %POS\n' "$PREFIX"_temp2.vcf | wc -l) \
	>> "$FILEDIR"/log.txt

##########
# filter sites with > 5% het & > 5% missing data
## calculate proportion het per site with plink
plink2 --vcf "$PREFIX"_temp2.vcf \
	--remove "$FILEDIR"/Cg.txt \
	--geno-counts cols=chrom,pos,ref,alt,homref,refalt,homalt1 \
	--allow-extra-chr \
	--double-id \
	--out "$PREFIX"

echo "plink genotype caluclation complete"

# Move frequency file to directory
mv "$PREFIX".gcount "$FILEDIR"/

## list of site ids for following Rscript
bcftools query -f'%CHROM %POS\n' "$PREFIX"_temp2.vcf > "$FILEDIR"/"$PREFIX"_sites.txt

## Produce list of sites to filter by allele frquency
Rscript "$FILEDIR"/007a_het_sites.R "$FILEDIR"/"$PREFIX".gcount 0.05 "$FILEDIR"/"$PREFIX"

# highly het sites to  remove file
cut -f1,2 "$FILEDIR"/"$PREFIX"_hetmin.txt | tail -n+2 > "$FILEDIR"/remove.txt

## remove highly heterozygous sites and sites with 5% missing calls
bcftools view --targets-file ^"$FILEDIR"/remove.txt "$PREFIX"_temp2.vcf -o "$PREFIX"_temp3.vcf

# log
echo "heterozygisty filter complete"

echo "Heterozygosity site filter" >> "$FILEDIR"/log.txt
echo $(bcftools query -f'%CHROM %POS\n' "$PREFIX"_temp3.vcf | wc -l) \
	>> "$FILEDIR"/log.txt

# calculate site missingness
plink2 --vcf "$PREFIX"_temp3.vcf --missing variant-only vcols=chrom,pos,nmiss,nobs,fmiss \
  --allow-extra-chr \
  --double-id \
  --out "$PREFIX"

# Keep sites where at least 95% is not missing
bcftools view --include 'F_MISSING<0.05' "$PREFIX"_temp3.vcf -Oz -o "$PREFIX"_filter2.vcf.gz
#bcftools view --include 'F_MISSING<0.2' "$PREFIX"_temp3.vcf -Oz -o "$PREFIX"_filter2.vcf.gz

echo "missing data filter complete"

echo "missing site call filter" >> "$FILEDIR"/log.txt
echo $(bcftools query -f'%CHROM %POS\n' "$PREFIX"_filter2.vcf.gz | wc -l) \
	>> "$FILEDIR"/log.txt

##########
# top cut depth at value of > Q3 + 1.5IQR
## calculate depth per site and plot
bcftools query -f '%CHROM %POS %DP\n' "$PREFIX"_filter2.vcf.gz \
	> "$FILEDIR"/depth.txt
Rscript "$FILEDIR"/007b_depth_thresh.R "$FILEDIR"/depth.txt "$FILEDIR"/"$PREFIX"

echo "depth calc complete"

## filter on depth
MAXDP=$(head -n1 "$FILEDIR"/"$PREFIX"_depth_cutval.txt)

bcftools view -i "INFO/DP < $MAXDP" "$PREFIX"_filter2.vcf.gz \
	-Oz -o "$PREFIX"_filter3.vcf.gz

## Write to filtering report
echo "Depth filter complete"
echo "depth filter " >> "$FILEDIR"/log.txt
echo $(bcftools query -f'%CHROM %POS\n' "$PREFIX"_filter3.vcf.gz | wc -l) \
	>> "$FILEDIR"/log.txt

##########
# calculate stats per ind
plink2 --vcf "$PREFIX"_filter3.vcf.gz \
	--sample-counts cols=homref,het,homalt \
	--allow-extra-chr \
        --double-id \
        --out "$PREFIX"

# move to file directory
mv "$PREFIX".scount "$FILEDIR"/
