---
title: "Noodling on LD decay"
author: "Maya Wilson Brown"
date: "2024-10-16"
output: html_document
---


First, the script that generated the calculations of LD
based on population distinctions
```{r}
pops <- read.csv("~/Documents/PhD/Research/capsella_population_structure/cbp_pop_str.txt", sep = "\t", header=T)

nyc <- pops[which(str_detect(pops$vcf_sample_name, "L003")),]
easia <- pops[which(pops$k3population == "E_Asia"),]

write.table(nyc$vcf_sample_name, "~/Documents/PhD/Research/vcf_filtering/ld_decay/nyc_vcf_sample.txt",
            row.names = F,
            col.names = F,
            quote = F)

write.table(easia$vcf_sample_name, "~/Documents/PhD/Research/vcf_filtering/ld_decay/easia_vcf_sample.txt",
            row.names = F,
            col.names = F,
            quote = F)
```


```{bash, code = readLines("~/Documents/PhD/Research/vcf_filtering/ld_decay/noodling_LD_decay.sh"), eval = FALSE}
```

The actual calculation of linkage into averaged bins is done by the following python script.
```{bash, code = readLines("~/Documents/PhD/Research/vcf_filtering/ld_decay/ld_decay_calc.py"), eval = FALSE}
```

Then, to plot the output
```{r}
library(ggplot2)
# read in data
nyc_ld_bins <- read.csv("~/Documents/PhD/Research/vcf_filtering/ld_decay/nyc_chrAll.ld_decay_bins",
                    sep = "\t",
                    header = T)


# plot LD decay
ggplot(nyc_ld_bins, aes(distance, avg_R2), colour = chr) + geom_line() +
  xlab("Distance (bp)") + ylab(expression(italic(r)^2)) +
  geom_hline(yintercept = 0.1, color = "red") +
  facet_wrap(~chr) +
  theme_bw()

# which bins size before LD reaches backgound level of about 0.1
ggplot(subset(nyc_ld_bins, avg_R2 <= 0.1), aes(distance, avg_R2), colour = chr) + geom_line() +
  xlab("Distance (bp)") + ylab(expression(italic(r)^2)) +
  geom_vline(xintercept = 500000, color = "red") +
  facet_wrap(~chr) +
  theme_bw()

ggplot(nyc_ld_bins, aes(distance, avg_R2), colour = chr) + geom_line() +
  xlab("Distance (bp)") + ylab(expression(italic(r)^2)) +
  geom_vline(xintercept = 500000, color = "red") +
  geom_hline(yintercept = 0.1, color = "blue") +
  facet_wrap(~chr) +
  theme_bw() +
  ggtitle("Linkage in New York City population")
```

Honestly, 500kb seems kind of long. I wonder if introgression from C. rubella might have something to do with it, so I will do the same analysis on the East Asian population.
```{r}
# read in data
ld_bins <- read.csv("~/Documents/PhD/Research/vcf_filtering/ld_decay/easia_chrAll.ld_decay_bins",
                    sep = "\t",
                    header = T)


# plot LD decay
ggplot(ld_bins, aes(distance, avg_R2), colour = chr) + geom_line() +
  xlab("Distance (bp)") + ylab(expression(italic(r)^2)) +
  geom_hline(yintercept = 0.1, color = "red") +
  facet_wrap(~chr) +
  theme_bw()

# which bins size before LD reaches backgound level of about 0.1
ggplot(subset(ld_bins, avg_R2 <= 0.1), aes(distance, avg_R2), colour = chr) + geom_line() +
  xlab("Distance (bp)") + ylab(expression(italic(r)^2)) +
  geom_vline(xintercept = 500000, color = "red") +
  facet_wrap(~chr) +
  theme_bw()

ggplot(ld_bins, aes(distance, avg_R2), colour = chr) + geom_line() +
  xlab("Distance (bp)") + ylab(expression(italic(r)^2)) +
  geom_vline(xintercept = 500000, color = "red") +
  geom_hline(yintercept = 0.1, color = "blue") +
  facet_wrap(~chr) +
  theme_bw()
```

This seems to indicate there is rarely LD oover 0.1 at any distance, which is equally confusing. I guess this just means I should use windows over 500kb?

I also did not exclude SNPs in centromeric regions so it may be worth making a linkage map too.

Another analysis, this time with the global sample of CBP and the variant only VCF
```{r}
# read in data
ld_bins <- read.csv("~/Documents/PhD/Research/vcf_filtering/ld_decay/var_cbp_JLv4_LD_chrAll.ld_decay_bins",
                    sep = "\t",
                    header = T)


# plot LD decay
ggplot(ld_bins, aes(distance, avg_R2), colour = chr) + geom_line() +
  xlab("Distance (bp)") + ylab(expression(italic(r)^2)) +
  geom_hline(yintercept = 0.1, color = "red") +
  facet_wrap(~chr) +
  theme_bw()

# which bins size before LD reaches backgound level of about 0.1
ggplot(subset(ld_bins, avg_R2 <= 0.1), aes(distance, avg_R2), colour = chr) + geom_line() +
  xlab("Distance (bp)") + ylab(expression(italic(r)^2)) +
  geom_vline(xintercept = 500000, color = "red") +
  facet_wrap(~chr) +
  theme_bw()

ggplot(ld_bins, aes(distance, avg_R2), colour = chr) + geom_line() +
  xlab("Distance (bp)") + ylab(expression(italic(r)^2)) +
  geom_vline(xintercept = 250000, color = "red") +
  geom_hline(yintercept = 0.1, color = "blue") +
  facet_wrap(~chr) +
  theme_bw()
```
250kb to get to a background LD of <0.2 (closer to about 0.1 which is the blue line plotted) is similar to *Arabidopsis thaliana*. On some chromosomes, there does seem to be long range LD starting at a distance of about 750kb, and I do believe **that** is due to introgression.

For some reason, I lose a LOT of variants during the admixtyure analysis so I am checking LD once again but usuing the VCF I generated after filtering for minor allele frquency in the species separately, and remove the flags for maf, geno and mind.
```{r}
# read in data
ld_bins <- read.csv("~/Documents/PhD/Research/vcf_filtering/ld_decay/var_cbp_JLv4_LD_maf_chrAll.ld_decay_bins",
                    sep = "\t",
                    header = T)


# plot LD decay
ggplot(ld_bins, aes(distance, avg_R2), colour = chr) + geom_line() +
  xlab("Distance (bp)") + ylab(expression(italic(r)^2)) +
  geom_hline(yintercept = 0.1, color = "red") +
  facet_wrap(~chr) +
  theme_bw()

# which bins size before LD reaches backgound level of about 0.1
ggplot(subset(ld_bins, avg_R2 <= 0.1), aes(distance, avg_R2), colour = chr) + geom_line() +
  xlab("Distance (bp)") + ylab(expression(italic(r)^2)) +
  geom_vline(xintercept = 500000, color = "red") +
  facet_wrap(~chr) +
  theme_bw()

ggplot(ld_bins, aes(distance, avg_R2), colour = chr) + geom_line() +
  xlab("Distance (bp)") + ylab(expression(italic(r)^2)) +
  geom_vline(xintercept = 250000, color = "red") +
  geom_hline(yintercept = 0.1, color = "blue") +
  facet_wrap(~chr) +
  theme_bw()
```
Oh, okay :) Yeah, that is why.

Just to start looking at allele frequency.
```{r}
# I ran the following command with the same variables described in capsella_population_structure/runConditionalAdmixture.sh

#plink2 --vcf $VCF --allow-extra-chr --keep $ALL_CBP --set-all-var-ids @:# --freq --out CBP_capsella

afreq <- read.delim("~/Desktop/CBP_capsella.afreq")

# Plot frquencies
ggplot() + geom_histogram(data = afreq, aes(ALT_FREQS), binwidth = 0.02)

# Plot frequencies ater filtering MAF > 0.05
ggplot() + geom_histogram(data = subset(afreq, ALT_FREQS > 0.05 & ALT_FREQS < 0.95), aes(ALT_FREQS), binwidth = 0.02)
```


I can just document running ReLERNN here, for now.
```{bash, eval = F}
# needs tensorflow and CUDA with CuDNN
module load TensorFlow/2.13.0-foss-2023a
module load  cuDNN/8.9.7.29-CUDA-12.3.0
ml scikit-learn/1.3.1-gfbf-2023a  matplotlib/3.7.2-gfbf-2023a 
# also pip installed scikit-allel and msprime

# I get some warnings: The scripts msp and mspms are installed in '/mnt/home/wils1582/.local/bin'.Consider adding this directory to PATH
# So I added it to my path like this PATH=/mnt/home/wils1582/.local/bin:$PATH
```


```{bash, eval = F}
#VCF must be biallelic with NO missing data, which is kind of kooky
ReLERNN_SIMULATE 
```



```

